receiver/kubernetesclusterreceiver/collector.go

```diff
@@ -0,0 +1,213 @@
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kubernetesclusterreceiver/utils"
+)
+
+// TODO: Consider moving some of these constants to
+// https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/conventions/opentelemetry.go.
+
+// Resource label keys.
+const (
+	// Resource labels keys for UID.
+	k8sKeyPodUID                   = "k8s.pod.uid"
+	k8sKeyNodeUID                  = "k8s.node.uid"
+	k8sKeyCronJobUID               = "k8s.cronjob.uid"
+	k8sKeyDeploymentUID            = "k8s.deployment.uid"
+	k8sKeyJobUID                   = "k8s.job.uid"
+	k8sKeyNamespaceUID             = "k8s.namespace.uid"
+	k8sKeyReplicaSetUID            = "k8s.replicaset.uid"
+	k8sKeyReplicationControllerUID = "k8s.replicationcontroller.uid"
+	k8sKeyStatefulSetUID           = "k8s.statefulset.uid"
+	k8sKeyDaemonSetUID             = "k8s.daemonset.uid"
+	k8sKeyHPAUID                   = "k8s.hpa.uid"
+	k8sKeyResourceQuotaUID         = "k8s.resourcequota.uid"
+
+	// Resource labels keys for Name.
+	k8sKeyCronJobName               = "k8s.cronjob.name"
+	k8sKeyNodeName                  = "k8s.node.name"
+	k8sKeyDeploymentName            = "k8s.deployment.name"
+	k8sKeyJobName                   = "k8s.job.name"
+	k8sKeyReplicaSetName            = "k8s.replicaset.name"
+	k8sKeyReplicationControllerName = "k8s.replicationcontroller.name"
+	k8sKeyStatefulSetName           = "k8s.statefulset.name"
+	k8sKeyDaemonSetName             = "k8s.daemonset.name"
+	k8sKeyHPAName                   = "k8s.hpa.name"
+	k8sKeyResourceQuotaName         = "k8s.resourcequota.name"
+
+	// Resource labels for container.
+	containerKeyID       = "container.id"
+	containerKeySpecName = "container.spec.name"
+)
+
+// dataCollector wraps around a metricsStore and a metadaStore exposing
+// methods to perform on the underlying stores
+type dataCollector struct {
+	logger                 *zap.Logger
+	metricsStore           *metricsStore
+	metadataStore          *metadataStore
+	nodeConditionsToReport []string
+}
+
+// newDataCollector returns a dataCollector.
+func newDataCollector(logger *zap.Logger, nodeConditionsToReport []string) *dataCollector {
+	return &dataCollector{
+		logger: logger,
+		metricsStore: &metricsStore{
+			metricsCache: make(map[types.UID][]consumerdata.MetricsData),
+		},
+		metadataStore: &metadataStore{
+			stores: map[string]cache.Store{},
+		},
+		nodeConditionsToReport: nodeConditionsToReport,
+	}
+}
+
+// setupMetadataStore initializes a metadata store for the kubernetes object.
+func (dc *dataCollector) setupMetadataStore(o runtime.Object, store cache.Store) {
+	dc.metadataStore.setupStore(o, store)
+}
+
+func (dc *dataCollector) removeFromMetricsStore(obj interface{}) {
+	key, err := utils.GetUIDForObject(obj.(runtime.Object))
+
+	if err != nil {
+		dc.logger.Error(
+			"failed to remove from metric cache",
+			zap.String("obj", reflect.TypeOf(obj).String()),
+			zap.Error(err),
+		)
+		return
+	}
+
```

Comment:
Should this logic maybe live inside the metric store?

Reply:
Moved this logic into `metricStore` for update and remove methods.